[feat] Add tooltip support for Vue nodes

christian-byrne opened the pull request:
## Summary

Added tooltip support for Vue node components using PrimeVue's v-tooltip directive with proper data integration and container scoping.

https://github.com/user-attachments/assets/d1af31e6-ef6a-4df8-8de4-5098aa4490a1

## Changes

- **What**: Implemented tooltip functionality for Vue node headers, input/output slots, and widgets using [PrimeVue v-tooltip](https://primevue.org/tooltip/) directive
- **Dependencies**: Leverages existing PrimeVue tooltip system, no new dependencies

## Review Focus

Container scoping implementation via provide/inject pattern for tooltip positioning, proper TypeScript interfaces eliminating `as any` casts, and integration with existing settings store for tooltip delays and enable/disable functionality.

```mermaid
graph TD
    A[LGraphNode Container] --> B[provide tooltipContainer]
    B --> C[NodeHeader inject]
    B --> D[InputSlot inject]
    B --> E[OutputSlot inject]
    B --> F[NodeWidgets inject]

    G[useNodeTooltips composable] --> H[NodeDefStore lookup]
    G --> I[Settings integration]
    G --> J[i18n fallback]

    C --> G
    D --> G
    E --> G
    F --> G

    style A fill:#f9f9f9,stroke:#333,color:#000
    style G fill:#e8f4fd,stroke:#0066cc,color:#000
```

christian-byrne committed:
[feat] Add tooltip support for Vue nodes

Implements comprehensive tooltip functionality for all Vue node components:
- Node headers display node descriptions from definitions
- Input/output slots show slot-specific tooltip information
- Widgets display tooltip text from metadata and node definitions
- Uses PrimeVue's v-tooltip directive with container scoping
- Integrates with existing settings (Comfy.EnableTooltips, LiteGraph.Node.TooltipDelay)
- Provides fallback content when no tooltip data is available
- Full TypeScript support with proper interfaces

christian-byrne committed:
Update tooltip positioning to use left/right instead of top/bottom

- Input slots and widgets: show tooltips on right (auto-fallback to left)
- Output slots: show tooltips on left (auto-fallback to right)
- Improves UX by not covering connection areas and using horizontal space better

christian-byrne committed:
Apply Figma design styling to tooltips using PrimeVue passthrough API

- Use Tailwind classes via pt.text and pt.arrow for clean styling
- Match Figma specs: dark background (#171718), purple border (#5B5E7D)
- Fix tooltip positioning: inputs show left, outputs show right
- Proper arrow styling with before:border-[#171718] for chevron shape

AustinMroz reviewed (dismissed):
LGTM

DrJKL reviewed (dismissed):
👍🏻

christian-byrne committed:
[fix] use proper generic type for settings store mock - addresses review feedback

Co-authored-by: DrJKL <DrJKL@users.noreply.github.com>

christian-byrne committed:
[fix] replace any types with proper ComfyNodeDef structure - addresses review feedback

Co-authored-by: DrJKL <DrJKL@users.noreply.github.com>

christian-byrne committed:
[fix] remove .left modifier from tooltip to allow auto-positioning near cursor - addresses review feedback

Co-authored-by: DrJKL <DrJKL@users.noreply.github.com>

christian-byrne committed:
[style] replace arbitrary color values with theme colors - addresses review feedback

Co-authored-by: DrJKL <DrJKL@users.noreply.github.com>

christian-byrne commented on src/renderer/extensions/vueNodes/components/NodeHeader.spec.ts:
```diff
@@ -24,19 +26,67 @@ const makeNodeData = (overrides: Partial<VueNodeData> = {}): VueNodeData => ({
+  const nodeDefStore = useNodeDefStore()
+
+  // Mock tooltip delay setting
+  vi.spyOn(settingStore, 'get').mockImplementation((key) => {
```
Fixed in 57b5429 - now properly typed with generic Settings constraint

christian-byrne committed:
[revert] restore .left tooltip positioning - keep design intent of tooltips outside node boundaries

Co-authored-by: DrJKL <DrJKL@users.noreply.github.com>

christian-byrne commented on src/renderer/extensions/vueNodes/composables/useNodeTooltips.ts:
```diff
@@ -0,0 +1,120 @@
+      pt: {
+        text: {
+          class:
+            'bg-[#171718] border border-[#5B5E7D] rounded-md px-4 py-2 text-white text-sm font-normal leading-tight max-w-[300px] shadow-none'
```
Fixed in 5b672cb - replaced arbitrary hex values with theme colors (bg-charcoal-100, border-slate-300)

christian-byrne committed:
[fix] update settingStore import path after rebase

DrJKL reviewed (dismissed):
Not sure if the charcoal colors need to be flipped after https://github.com/Comfy-Org/ComfyUI_frontend/pull/5609.

Nothing critical, looks good 👍🏻

christian-byrne committed:
[refactor] destructure props in NodeHeader for reactivity - addresses review feedback

Co-authored-by: DrJKL <DrJKL@users.noreply.github.com>

christian-byrne committed:
[refactor] destructure props with default in NodeSlots - addresses review feedback

Co-authored-by: DrJKL <DrJKL@users.noreply.github.com>

christian-byrne committed:
Use MaybeRef for nodeType parameter in useNodeTooltips

Allow passing either string or ref for more flexible API usage.

Co-Authored-By: DrJKL <DrJKL@users.noreply.github.com>

christian-byrne committed:
Replace max-w-[300px] with max-w-75 in tooltip styling

Use standard Tailwind class instead of arbitrary value.

Co-Authored-By: DrJKL <DrJKL@users.noreply.github.com>

christian-byrne commented:
✅ **Comment #6 resolved** - Destructured props with default in NodeSlots.vue

Fixed in a03816bb72c4eb7b2055e6e54b55fde85e78e7b9 - added default value for nodeData prop and removed unused lodLevel parameter.

christian-byrne commented:
✅ **Comment #7 resolved** - Destructured props in NodeWidgets.vue

Fixed in c0756aa6ad8c6077ad88e9be1e99598e8e10dc73 - replaced computed wrapper with direct prop access, maintaining Vue reactivity patterns.

christian-byrne commented:
✅ **Comment #8 resolved** - Used MaybeRef for nodeType parameter

Fixed in b85e8f1b40bb9b7b1b1b8e2c41ed4ba9b0e7e42c - updated useNodeTooltips composable to accept MaybeRef<string> for more flexible API usage, using unref() for reactive unwrapping.

christian-byrne commented:
✅ **Comment #9 resolved** - Replaced max-w-[300px] with max-w-75

Fixed in 951f29a97ff8f24f8b5651a0b65cd7fa1b9b0f3b - updated tooltip styling to use standard Tailwind class instead of arbitrary value.

christian-byrne commented on src/renderer/extensions/vueNodes/components/NodeHeader.vue:
```diff
@@ -72,9 +75,24 @@ onErrorCaptured((error) => {
 const isEditing = ref(false)
 
-const nodeInfo = computed(() => props.nodeData || props.node)
+const nodeInfo = computed(() => props.nodeData)
```
✅ **Fixed in a8b610c** - Destructured props directly: `const { nodeData, readonly, collapsed } = defineProps<NodeHeaderProps>()` - removed computed wrapper for better Vue 3 reactivity patterns.

christian-byrne commented on src/renderer/extensions/vueNodes/components/NodeSlots.vue:
```diff
@@ -32,23 +34,22 @@ import { computed, onErrorCaptured, ref } from 'vue'
 const props = defineProps<NodeSlotsProps>()
 
-const nodeInfo = computed(() => props.nodeData || props.node || null)
+const nodeInfo = computed(() => props.nodeData || null)
```
✅ **Fixed in b975601** - Destructured props with default: `const { nodeData = null, readonly } = defineProps<NodeSlotsProps>()` - removed unused lodLevel and computed wrapper.

christian-byrne commented on src/renderer/extensions/vueNodes/components/NodeWidgets.vue:
```diff
@@ -78,7 +78,15 @@ onErrorCaptured((error) => {
 })
 
-const nodeInfo = computed(() => props.nodeData || props.node)
+const nodeInfo = computed(() => props.nodeData)
```
✅ **Fixed in 6b3ed01** - Destructured props: `const { nodeData, readonly, lodLevel } = defineProps<NodeWidgetsProps>()` - replaced computed wrapper with direct prop access for better Vue 3 patterns.

christian-byrne commented on src/renderer/extensions/vueNodes/composables/useNodeTooltips.ts:
```diff
@@ -0,0 +1,120 @@
+ */
+export function useNodeTooltips(
+  nodeType: string,
+  containerRef?: Ref<HTMLElement | undefined>
```
✅ **Fixed in b6a976b** - Updated to use `MaybeRef<string>` for nodeType parameter and `unref()` for reactive unwrapping - allows passing either string or ref for more flexible API usage.

christian-byrne commented on src/renderer/extensions/vueNodes/composables/useNodeTooltips.ts:
```diff
@@ -0,0 +1,120 @@
+      pt: {
+        text: {
+          class:
+            'bg-charcoal-100 border border-slate-300 rounded-md px-4 py-2 text-white text-sm font-normal leading-tight max-w-[300px] shadow-none'
```
Thanks for the clarification! You were right - `max-w-75` (300px equivalent) works perfectly. Much cleaner than the arbitrary value.

actions-user committed:
[auto-fix] Apply ESLint and Prettier fixes
